src/plugins/dnf/subscription_manager.py

```diff
@@ -241,7 +244,46 @@ def transaction(self):
         cfg = config.get_config_parser()
         if "1" == cfg.get("rhsm", "package_profile_on_trans"):
             log.debug("Uploading package profile")
-            package_profile_client = ProfileActionClient()
-            package_profile_client.update()
+            self._upload_profile()
         else:
             log.debug("Uploading package profile disabled in configuration file")
+
+    def _upload_profile_blocking(self) -> None:
+        """
+        Try to upload DNF profile to server
+        """
+        with LiveStatusMessage("Uploading DNF profile"):
+            try:
+                profile_mgr = inj.require(inj.PROFILE_MANAGER)
+                identity = inj.require(inj.IDENTITY)
+                profile_mgr.update_check(self.cp, identity.uuid)
+            except RemoteServerException as err:
+                # When it is not possible to upload profile ATM, then print only error about this
+                # to rhsm.log. The rhsmcertd will try to upload it next time.
+                log.error("Unable to upload profile: {err}".format(err=str(err)))
+
+    def _upload_profile(self) -> None:
+        """
+        Try to upload DNF profile to server, when it is supported by server. This method
+        tries to "outsource" this activity to rhsmcertd first. When it is not possible due to
+        various reasons, then we try to do it ourselves in blocking way.
+        """
+        # First try to get PID of rhsmcertd from lock file
+        try:
+            with open("/var/lock/subsys/rhsmcertd", "r") as lock_file:
+                rhsmcertd_pid = int(lock_file.readline())
+        except (IOError, ValueError) as err:
+            log.info(f"Unable to read rhsmcertd lock file: {err}")
+        else:
+            if is_process_running("rhsmcertd", rhsmcertd_pid) is True:
+                # This will only send SIGUSR1 signal, which triggers gathering and uploading
+                # of DNF profile by rhsmcertd. We try to "outsource" this activity to rhsmcertd
+                # server to not block registration process
+                log.debug("Sending SIGUSR1 signal to rhsmcertd process")
+                try:
+                    os.kill(rhsmcertd_pid, signal.SIGUSR1)
+                except OSError as err:
+                    log.warning(f"Unable to send SIGUSR1 signal to rhsmcertd process ({rhsmcertd_pid}): {err}")
```

Comment:
hm this is `log.debug()` during the registration; should the same be used here for consistency?

also, after this log, most likely the blocking upload ought to be called:
```python
                    self._upload_profile_blocking()
```

Reply:
The `self._upload_profile_blocking()` will not be called (it is in different block)... I will add this call to this block too.

When it is not possible to send signal, then it could have some serious reason (not enough permissions, SELinux issue, etc.) or there is some edge cases (rhsmcertd recently died, etc.), but it will be hopefully solved by sending DNF profile in blocking. If sending in blocking mode will fail, then error will be written to log file. Thus, it is question if it is good to decrease severity of this log message.

Reply:
> Thus, it is question if it is good to decrease severity of this log message.

I'm generally fine with warning as well, with a bit of worry that there will be a "warning" that will be noticed in logs.

In case it is left as warning, please adapt the same code in the registration (`src/subscription_manager/cli_command/register.py`, `_upload_profile()`) to match this for consistency.